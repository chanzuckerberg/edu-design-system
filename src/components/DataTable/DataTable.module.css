@import '../../design-tokens/mixins.css';

/*------------------------------------*\
    # DATA TABLE
\*------------------------------------*/

/**
 * Wraps the label and the optional filter button
 * 1) Vertically align items in this row at the bottom; this is only needed
 *    when the label text breaks to multiple rows; the filter button will
 *    remain aligned to the bottom of this row
 * 2) If filter button is present, position it at the end of the flex row
 */
.data-table__label-wrapper {
  display: flex;
  align-items: flex-end;
  margin-bottom: var(--eds-size-2);

  .data-table--filters & {
    justify-content: space-between; /* 2 */
  }
}

/**
 * The label title above the data table provides an accessible name.
 */
.data-table__label {
  @mixin eds-typography-preset-003;
  color: var(--eds-theme-color-text-neutral-strong);
}


