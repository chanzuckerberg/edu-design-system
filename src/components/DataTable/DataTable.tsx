import clsx from 'clsx';
import React, { createContext } from 'react';
import { useUID } from 'react-uid';
import styles from './DataTable.module.css';
import Button from '../Button';
import DataTableColumn from '../DataTableColumn';
import Icon from '../Icon';
import { Table, Props as TableProps } from '../Table/Table';

export interface Props extends TableProps {
  /**
   * Each column has a title, a selector, and optional data.
   * 1) The title is the text that will be displayed in the table header cell
   * 2) The selector identifies which row property belongs to
   * each column; ex. column.selector = 'name' refers to () row => row.name
   * 3) Data (optional) can be used to store an array of values from rows
   */
  columns: DataTableColumnType[];
  /**
   * HTML id attribute for accessibility.
   */
  id?: string;
  /**
   * Accessible label to associate title and table
   */
  label?: string;
  /**
   * Each cell in a row contains a value that is also associated with a cell in columns.
   */
  rows: DataTableRowType;
  /**
   * Exposes a button for filtering options
   */
  useFilters?: boolean;
}

/**
 * 1) Each item in the data array needs to be in the format of key:string/value:string pairs
 */
export type DataTableColumnType = {
  data?: Record<string, string>[] /* 1 */;
  selector: string;
  title: string;
};

/**
 * A DataTable row needs to be in the format of an array of key:string/value:string pairs
 */
export type DataTableRowType = Record<string, string>[];

export type DataTableContextType = {
  columns?: DataTableColumnType[];
  rows?: DataTableRowType;
};

/**
 * DataTableContext.Provider wraps the DataTable and can be used to store context from props or state, such as table data.
 */
export const DataTableContext = createContext<DataTableContextType>({});

/**
 * BETA: This component is still a work in progress and is subject to change.
 *
 * ```ts
 * import {DataTable} from "@chanzuckerberg/eds";
 * ```
 *
 * DataTable component extends the functionality of Table component
 * by accepting structured data rows and columns, which can then be
 * filtered, sorted, selected, etc.
 */
export const DataTable = ({
  className,
  caption,
  columns,
  id,
  label,
  rows,
  useFilters = false,
  ...other
}: Props) => {
  const componentClassName = clsx(
    styles['data-table'],
    useFilters === true && styles['data-table--filters'],
    className,
  );

  const generatedId = useUID();
  const dataTableId = id || generatedId;

  return (
    <DataTableContext.Provider value={{ columns, rows }}>
      <div className={componentClassName}>
        <div className={styles['data-table__label-wrapper']}>
          <label className={styles['data-table__label']} htmlFor={id}>
            {label}
          </label>
          {useFilters && (
            <Button size="md" status="brand" variant="secondary">
              <Icon name="filter-list" purpose="decorative" size="1.375rem" />
              Filters
            </Button>
          )}
        </div>
        <Table
          caption={caption}
          className={styles['data-table__table']}
          id={dataTableId}
          {...other}
        >
          <Table.Header>
            <Table.Row className={styles['data-table__table']}>
              {columns.map((col, index) => (
                <DataTableColumn
                  className={styles['data-table__header-cell']}
                  key={index}
                  selector={col.selector}
                  title={col.title}
                />
              ))}
            </Table.Row>
          </Table.Header>
          <Table.Body>
            {rows.map((row, rowIndex) => (
              <Table.Row key={rowIndex}>
                {columns.map((col, colIndex) => (
                  <Table.Cell
                    data-heading={col.title}
                    id={`row${colIndex}-${col.selector}`}
                    key={colIndex}
                  >
                    {rows[rowIndex][col.selector]}
                  </Table.Cell>
                ))}
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
      </div>
    </DataTableContext.Provider>
  );
};
