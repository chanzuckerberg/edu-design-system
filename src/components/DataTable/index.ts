export { DataTable as default } from './DataTable';
export type {
  DataTableProps,
  DataTableRowProps,
  DataTableWithStatus,
  DataTableSortDirection,
} from './DataTable';

// Re-export Tanstack hooks and functions to consumers
// https://tanstack.com/table/latest
export * as utils from '@tanstack/react-table';
