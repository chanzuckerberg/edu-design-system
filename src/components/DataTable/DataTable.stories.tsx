import { BADGE } from '@geometricpanda/storybook-addon-badges';
import { StoryObj, Meta } from '@storybook/react';
import React from 'react';
import { DataTable } from './DataTable';

export default {
  title: 'Organisms/Tables/DataTable',
  component: DataTable,
  parameters: {
    badges: [BADGE.BETA],
  },
  decorators: [
    (Story) => (
      <div
        style={{
          margin: '0.25rem',
        }}
      >
        <Story />
      </div>
    ),
  ],
} as Meta<Args>;

type Args = React.ComponentProps<typeof DataTable>;

const tableColumns = [
  {
    title: 'Student',
    selector: 'student',
  },
  {
    title: 'Status',
    selector: 'status',
  },
  {
    title: 'Updated',
    selector: 'updated',
  },
  {
    title: 'Completed',
    selector: 'completed',
  },
];

const tableRows = [
  {
    student: 'Flores, Juanita',
    status: 'Needs Feedback',
    updated: '5/1/2022',
    completed: '50%',
  },
  {
    student: 'Miles, Esther',
    status: 'Stop and Revise',
    updated: '6/13/2022',
    completed: '33%',
  },
  {
    student: 'Black, Marvin',
    status: 'Finish Work',
    updated: '4/12/2022',
    completed: '20%',
  },
  {
    student: 'Nguyen, Savannah',
    status: 'Needs Feedback',
    updated: '12/9/2021',
    completed: '80%',
  },
  {
    student: 'Alexander, Leslie',
    status: 'Continue Working',
    updated: '5/29/2022',
    completed: '20%',
  },
  {
    student: 'Bell, Jerome',
    status: 'Review Feedback',
    updated: '2/21/2022',
    completed: '5%',
  },
  {
    student: 'Richards, Ronald',
    status: 'Needs Feedback',
    updated: '6/28/2022',
    completed: '60%',
  },
  {
    student: 'Simmons, Brooklyn',
    status: 'Stop and Revise',
    updated: '4/18/2022',
    completed: '40%',
  },
  {
    student: 'Williamson, Cameron',
    status: 'Finish Work',
    updated: '5/14/2022',
    completed: '50%',
  },
];

export const Default: StoryObj<Args> = {
  args: {
    caption: 'This is a dataTable caption and it is required',
    columns: tableColumns,
    label: 'Data Table',
    rows: tableRows,
    useFilters: false,
    variant: 'zebra',
  },
};

export const Filters: StoryObj<Args> = {
  args: {
    ...Default.args,
    useFilters: true,
  },
};
