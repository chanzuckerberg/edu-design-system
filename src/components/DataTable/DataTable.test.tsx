import { generateSnapshots } from '@chanzuckerberg/story-utils';
import * as stories from './DataTable.stories';

describe('<DataTable />', () => {
  generateSnapshots(stories);
});
