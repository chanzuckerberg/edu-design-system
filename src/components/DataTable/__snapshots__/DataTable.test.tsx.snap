// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTable /> Default story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <div
    class="data-table"
  >
    <div
      class="data-table__label-wrapper"
    >
      <label
        class="data-table__label"
        for="1"
      >
        Data Table
      </label>
    </div>
    <table
      class="table data-table__table table--zebra"
      id="1"
    >
      <caption
        class="table__caption u-is-vishidden"
      >
        This is a dataTable caption and it is required
      </caption>
      <thead
        class="table__header"
      >
        <tr
          class="table__row data-table__table"
        >
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="student"
          >
            Student
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="status"
          >
            Status
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="updated"
          >
            Updated
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="completed"
          >
            Completed
          </th>
        </tr>
      </thead>
      <tbody
        class="table__body"
      >
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Flores, Juanita
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/1/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            50%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Miles, Esther
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Stop and Revise
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            6/13/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            33%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Black, Marvin
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Finish Work
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            4/12/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            20%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Nguyen, Savannah
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            12/9/2021
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            80%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Alexander, Leslie
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Continue Working
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/29/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            20%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Bell, Jerome
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Review Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            2/21/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            5%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Richards, Ronald
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            6/28/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            60%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Simmons, Brooklyn
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Stop and Revise
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            4/18/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            40%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Williamson, Cameron
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Finish Work
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/14/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            50%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<DataTable /> Filters story renders snapshot 1`] = `
<div
  style="margin: 0.25rem;"
>
  <div
    class="data-table data-table--filters"
  >
    <div
      class="data-table__label-wrapper"
    >
      <label
        class="data-table__label"
        for="3"
      >
        Data Table
      </label>
      <button
        class="clickable-style button button--secondary clickable-style--md clickable-style--secondary clickable-style--brand"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="icon"
          fill="currentColor"
          height="1.375rem"
          style="--icon-size: 1.375rem;"
          width="1.375rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <use
            xlink:href="test-file-stub#filter-list"
          />
        </svg>
        Filters
      </button>
    </div>
    <table
      class="table data-table__table table--zebra"
      id="3"
    >
      <caption
        class="table__caption u-is-vishidden"
      >
        This is a dataTable caption and it is required
      </caption>
      <thead
        class="table__header"
      >
        <tr
          class="table__row data-table__table"
        >
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="student"
          >
            Student
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="status"
          >
            Status
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="updated"
          >
            Updated
          </th>
          <th
            class="table__header-cell data-table__column data-table__header-cell table-header-cell--none"
            selector="completed"
          >
            Completed
          </th>
        </tr>
      </thead>
      <tbody
        class="table__body"
      >
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Flores, Juanita
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/1/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            50%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Miles, Esther
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Stop and Revise
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            6/13/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            33%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Black, Marvin
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Finish Work
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            4/12/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            20%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Nguyen, Savannah
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            12/9/2021
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            80%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Alexander, Leslie
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Continue Working
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/29/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            20%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Bell, Jerome
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Review Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            2/21/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            5%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Richards, Ronald
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Needs Feedback
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            6/28/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            60%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Simmons, Brooklyn
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Stop and Revise
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            4/18/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            40%
          </td>
        </tr>
        <tr
          class="table__row"
        >
          <td
            class="table__cell"
            data-heading="Student"
            id="row0-student"
          >
            Williamson, Cameron
          </td>
          <td
            class="table__cell"
            data-heading="Status"
            id="row1-status"
          >
            Finish Work
          </td>
          <td
            class="table__cell"
            data-heading="Updated"
            id="row2-updated"
          >
            5/14/2022
          </td>
          <td
            class="table__cell"
            data-heading="Completed"
            id="row3-completed"
          >
            50%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
